Feature/4.0.0

Timovzl opened the pull request:
- Generic JSON serializers instead of generated.

- Removed #if NET7/8 conditionals, now that 8 is the minimum version.

- Upgraded LangVersion and handled compiler suggestions.

- Suppressions and summary corrections.

- Implemented formatting/parsing via default interface implementations and (.NET 10+) extension members.
- Prepared outcommented precompiler directives for a .NET 10 patch upgrade, to reduce assembly bloat then.
- Also corrected how formattable/parsable is determined based on wrapped types.
- Also fixed a few bugs.

- Generator performance and cleanup.

- Added serialization to/from deepest underlying type (recursive).
- Also cleaned up the way in which lines not being source generated are outcommented.
- Also improved source generator performance.
- Also prevented transient issues when reporting warnings on wrapper types.

- Added wrapper EF collations, collation checks, and provider comparers.

Timovzl committed:
.NET 8+ only, IIdentity implements IWrapperValueObject, bug fixes, trimming, warnings.

Timovzl committed:
Generic JSON serializers instead of generated.

Timovzl committed:
Removed #if NET7/8 conditionals, now that 8 is the minimum version.

Timovzl committed:
Upgraded LangVersion and handled compiler suggestions.

Timovzl committed:
Suppressions and summary corrections.

Timovzl committed:
Implemented formatting/parsing via default interface implementations and (.NET 10+) extension members.

Prepared outcommented precompiler directives for a .NET 10 patch upgrade, to reduce assembly bloat then.

Also corrected how formattable/parsable is determined based on wrapped types.

Also fixed a few bugs.

Timovzl committed:
Removed outcommented code.

Timovzl committed:
Generator performance and cleanup.

Timovzl committed:
Added EnumerableComparer overloads that avoid boxing for ImmutableAray. (ArraySegment, the other common collection struct, has implicit conversions from arrays, which mess with nullability constraints.)

Timovzl committed:
Added serialization to/from deepest underlying type (recursive).

Also cleaned up the way in which lines not being source generated are outcommented.

Also improved source generator performance.

Also prevented transient issues when reporting warnings on wrapper types.

Timovzl committed:
Added wrapper EF collations, collation checks, and provider comparers.

Timovzl committed:
Shortened release notes

Timovzl committed:
Added [CompilerGenerated] to all generated types.

Timovzl committed:
Added Analyzer project, and analyzer for inadvertent comparisons based on implicit conversion.

Timovzl committed:
Moved Equals() and Compare() helpers from generated into helper class.

Timovzl committed:
Generated [Wrapper]ValueObjects can now be [structs and/or] records and/or use a custom base class.

Timovzl committed:
Nicer DebuggerDisplay of Wrappers/Identities.

Timovzl committed:
Made Entity.Id private init.

Timovzl committed:
DummyBuilder records clone themselves on each step, for reuse.

Timovzl committed:
DefinedEnum<TEnum> for validated enums.

Timovzl committed:
Revert "Made Entity.Id private init."

This reverts commit 8d46796ad5a9384470d73f6365fce6dfd0426cbc.

Timovzl committed:
Replaced DefinedEnums by enum validation.

JoergHoffmannatGitHub commented on DomainModeling.Generator/Configurators/EntityFrameworkConfigurationGenerator.cs:
```diff
@@ -400,15 +734,102 @@ public override InstantiationBinding With(IReadOnlyList<ParameterBinding> parame
 ";
 
-		AddSource(context, source, "EntityFrameworkDomainModelConfigurationExtensions", $"{Constants.DomainModelingNamespace}.EntityFramework");
+		AddSource(context, source, "EntityFrameworkDomainModelConfigurationExtensions", $"Architect.DomainModeling.EntityFramework");
```
error CS0016: Could not write to output file
I think the path is to long on my machine
I checked out to 'C:\Development\GitHub\Architect.DomainModeling'

Timovzl replied:
Hi @JoergHoffmannatGitHub, this works fine for me if I imitate your approach:

```
D:\Development\GitHub>git clone https://github.com/TheArchitectDev/Architect.DomainModeling.git
D:\Development\GitHub>cd Architect.DomainModeling
D:\Development\GitHub\Architect.DomainModeling>dotnet build
```

What is the reason you are trying to build from source, particularly in the context of this PR?

JoergHoffmannatGitHub replied:
You are right. Using 'dotnet build' works fine.
Personally, I prefer building with Visual Studio Community.

Timovzl replied:
Curious that the results differ. I get the same result (success) in VS.

Do you have the solution file under `C:\Development\GitHub\Architect.DomainModeling` or `C:\Development\GitHub\Architect.DomainModeling\Architect.DomainModeling`?

What are you trying to achieve by building, and what is the relation to this PR?

JoergHoffmannatGitHub replied:
The solution is located at C:\Development\GitHub\Architect.DomainModeling.
I am playing around with some ideas on how to use the attributes for defining a domain model to understand the generated code and how the entity, value object, and others work together.
The easiest way was to generate some classes in the sample project.
Ultimately, I plan to use the Nuget package.

Timovzl replied:
Understood. It's unclear to me why you're seeing the path length issue and I'm not, for the same IDE and local path.

For what it's worth, you might be able to do what you're trying to do with the NuGet package as well. Or you could use a shorter local path as an annoying workaround. 😛

Timovzl committed:
Entity == operator, entity ID generation type param moved from base to attribute, and attribute inheritance.

JoergHoffmannatGitHub commented:
The code-base that I am currently working with consists of a large set of binaries that are all signed. The savvy .NET devs out there will know that any assembly that’s used/referenced by a signed assembly must also be signed.

This is an issue when dealing with Architect.DomainModeling library that is not signed.

Code-Chops commented on DomainModeling.CodeFixProviders/EntityBaseClassWithIdTypeGenerationCodeFixProvider.cs:
```diff
@@ -0,0 +1,133 @@
+
+	public sealed override ImmutableArray<string> FixableDiagnosticIds => FixableDiagnosticIdConstant;
+
+	public sealed override FixAllProvider GetFixAllProvider()
```
redundant `sealed`. Changing this one to an expression bodied method is also more consistent

Code-Chops commented on DomainModeling.CodeFixProviders/EntityBaseClassWithIdTypeGenerationCodeFixProvider.cs:
```diff
@@ -0,0 +1,133 @@
+		return WellKnownFixAllProviders.BatchFixer;
+	}
+
+	public sealed override async Task RegisterCodeFixesAsync(CodeFixContext context)
```
redundant `sealed`

Code-Chops commented on DomainModeling.CodeFixProviders/EntityBaseClassWithIdTypeGenerationCodeFixProvider.cs:
```diff
@@ -0,0 +1,133 @@
+		if (root is null)
+			return;
+
+		if (root.FindNode(diagnostic.Location.SourceSpan) is not BaseTypeSyntax baseTypeSyntax)
```
Null propagation can be used here, so lines 31-32 can be removed

Code-Chops commented on DomainModeling.CodeFixProviders/EntityBaseClassWithIdTypeGenerationCodeFixProvider.cs:
```diff
@@ -0,0 +1,133 @@
+		if (baseTypeSyntax.Type is not GenericNameSyntax { Arity: 2, TypeArgumentList.Arguments.Count: 2, } entityBaseTypeSyntax)
+			return;
+		var semanticModel = await context.Document.GetSemanticModelAsync(context.CancellationToken).ConfigureAwait(false);
+		if (semanticModel.GetTypeInfo(baseTypeSyntax.Type, context.CancellationToken).Type is not
```
`entityBaseType` is never used

Code-Chops commented on DomainModeling.CodeFixProviders/EntityBaseClassWithIdTypeGenerationCodeFixProvider.cs:
```diff
@@ -0,0 +1,133 @@
+
+		var action = CodeAction.Create(
+			title: "Move type parameter for ID's underlying type into EntityAttribute",
+			createChangedDocument: ct => ConvertToAttributeAsync(context.Document, root, tds, entityBaseTypeSyntax),
```
`ct` could be a discard `_`

Code-Chops commented on DomainModeling.Generator/DomainEventGenerator.cs:
```diff
@@ -49,7 +49,7 @@ private static bool FilterSyntaxNode(SyntaxNode node, CancellationToken cancella
 
 		// Only with the attribute
-		if (type.GetAttribute("DomainEventAttribute", "Architect.DomainModeling", arity: 0) is null)
+		if (type.GetAttribute(attr => attr.IsOrInheritsClass("DomainEventAttribute", "Architect", "DomainModeling", arity: 0, out _)) is null)
```
There is commented out code on line 78

Code-Chops commented on DomainModeling.Generator/DummyBuilderGenerator.cs:
```diff
@@ -44,10 +44,10 @@ private static bool FilterSyntaxNode(SyntaxNode node, CancellationToken cancella
 
 		// Only with the attribute
-		if (type.GetAttribute("DummyBuilderAttribute", "Architect.DomainModeling", arity: 1) is not AttributeData { AttributeClass: not null } attribute)
+		if (type.GetAttribute(attr => attr.IsOrInheritsClass("DummyBuilderAttribute", "Architect", "DomainModeling", arity: 1, out _)) is not { } attribute)
```
You could change line 124 to: 
```
result.HasBuildMethod = members.Any(member => member.Name == "Build" && member is IMethodSymbol { Parameters.Length: 0 });
```

Code-Chops commented on DomainModeling.Generator/IdentityGenerator.cs:
```diff
@@ -186,28 +198,35 @@ private static bool FilterSyntaxNode(SyntaxNode node, CancellationToken cancella
+		var hasEntityAttribute = !hasIdentityAttribute && HasRelevantEntityAttribute(type, out attributeType);
+
+		// Path A (legacy): An Entity subclass that might be an Entity<TId, TUnderlying> for which TId may have to be generated
+		if (attributeType is null)
```
`Expression is always false according to nullable reference types' annotations`

Code-Chops commented on DomainModeling.Generator/TypeSymbolExtensions.cs:
unnecessary type check at line 24: `INamespaceSymbol`

Code-Chops commented on DomainModeling/Attributes/EntityAttribute.cs:
```diff
@@ -6,14 +6,38 @@ namespace Architect.DomainModeling;
+/// </summary>
+[AttributeUsage(AttributeTargets.Class, AllowMultiple = false, Inherited = false)]
+public class EntityAttribute : Attribute
+{
```
Body could be removed

Timovzl commented:
@JoergHoffmannatGitHub Could you open an issue for us to discuss the topic separately from this PR?

JoergHoffmannatGitHub commented:
Issue #27 is created.

Timovzl committed:
- Generated default ctor and deserialization now accounts for situations where a default ctor is infeasible.
- New analyzer warns if field initializers might be skipped on a [Wrapper]ValueObject.

Timovzl committed:
Added .NET 10 and fixed an issue with the new parsing/formatting extension members.

Timovzl committed:
ValueObject comparison operators disallow null.

Timovzl committed:
Also added IValueWrapper<TValue> interfaces (without TWrapper), which now handle the Serialize() and Value members.

Timovzl committed:
Fixed suggestions by new .NET analyzers.

Timovzl committed:
Fixed a bug in ValueObjectGenerator where the ToString() override was being enforced for records, even if one was provided manually.

Timovzl committed:
Reduced enum sizes in generators.

Timovzl committed:
Improved dummy builders with interfaces and better copyable model type names (via aliases).

Timovzl committed:
Made csproj LangVersions consistent and suppressed a needless suggestion.

Timovzl committed:
Fixed bugs in InternalEnumExtensions.

Timovzl committed:
Fixed some generator type conversion safety bugs, and upped package versions.

Timovzl committed:
Enabled struct ValueObjects.

Timovzl committed:
Improved wrapper check for existing parameterized ctor.

Timovzl committed:
Wrappers no longer generated parameterized ctor if any exist.

Timovzl committed:
DefinedEnum.ThrowUndefinedInput() now has return type, to facilitate use in expressions.

Timovzl committed:
Made DummyBuilders honor pre-existing DummyBuilders and use them.

Timovzl committed:
Prefixed each generated file with // <auto-generated/>.

Timovzl committed:
Disabled CS1591 in generated code.

Timovzl committed:
Fixed CS1591 suppression comments.

Timovzl committed:
WrapperVO's nullable ctor now also not generated if any manual parameterized ctors present.

Timovzl committed:
DefinedEnum.ThrowUndefinedInput() now returns Exception, so that it can be used in switch branch (with 'throw') irrespective of switch return type.
